src/main/java/life/qbic/omero/BasicOMEROClient.java

```diff
@@ -20,367 +22,407 @@
+  private HashMap<Long, Set<DatasetData>> datasetMap;
 
-        this.username = username;
-        this.password = password;
-        this.hostname = hostname;
-        this.port = port;
+  public BasicOMEROClient(String username, String password, String hostname, int port) {
 
-        this.sessionId = "";
+    this.username = username;
+    this.password = password;
+    this.hostname = hostname;
+    this.port = port;
 
-    }
+    this.sessionId = "";
 
-    private void connect(String hostname, int port, String username, String password) throws Exception {
+  }
 
-        this.gateway = new Gateway(new SimpleLogger());
-        LoginCredentials cred = new LoginCredentials();
-        cred.getServer().setHostname(hostname);
-
-        if (port > 0) {
-            cred.getServer().setPort(port);
-        }
+  private void connect(String hostname, int port, String username, String password)
+      throws Exception {
 
-        cred.getUser().setUsername(username);
-        cred.getUser().setPassword(password);
+    this.gateway = new Gateway(new SimpleLogger());
+    LoginCredentials cred = new LoginCredentials();
+    cred.getServer().setHostname(hostname);
 
-        ExperimenterData user = gateway.connect(cred);
-        this.ctx = new SecurityContext(user.getGroupId());
-
-        this.sessionId = gateway.getSessionId(user);
+    if (port > 0) {
+      cred.getServer().setPort(port);
     }
 
-    public void connect() {
-        try {
+    cred.getUser().setUsername(username);
+    cred.getUser().setPassword(password);
 
-            this.connect(this.hostname, this.port, this.username, this.password);
-        } catch (Exception e) {
-            System.out.println(e);
+    ExperimenterData user = gateway.connect(cred);
+    this.ctx = new SecurityContext(user.getGroupId());
 
-            System.out.println("++++++++++++++++++++++++++++++++++++++");
-            System.out.println(e.getLocalizedMessage());
-            System.out.println(e.getMessage());
-            System.out.println(e.toString());
+    this.sessionId = gateway.getSessionId(user);
+  }
 
-        }
+  public void connect() {
+    try {
 
-    }
+      this.connect(this.hostname, this.port, this.username, this.password);
+    } catch (Exception e) {
+      System.out.println(e);
 
-    public String getSessionId(){
+      System.out.println("++++++++++++++++++++++++++++++++++++++");
+      System.out.println(e.getLocalizedMessage());
+      System.out.println(e.getMessage());
+      System.out.println(e.toString());
 
-        return this.sessionId;
     }
 
-    public void disconnect() {
-        this.gateway.disconnect();
+  }
+
+  public String getSessionId() {
+
+    return this.sessionId;
+  }
+
+  public void disconnect() {
+    this.gateway.disconnect();
+  }
+
+  /**
+   * Tries to build an image download link for a given imageID. null will be returned if the image
+   * can not be downloaded due to its format
+   * 
```

Comment:
Maybe we should return an empty String instead? I am not sure about the default behaviour for non-existing images.

Reply:
or better throw a NullPointerException since it is thrown when the result is `null`

Reply:
d453508